### PR TITLE
DellEMC: Z9332f SFP enhancements

### PR DESCRIPTION
**Pull request by aravindmani-1**
#### Why I did it
400G media EEPROM and DOM information are not populated properly in DellEMC Z9332f platform.
#### How I did it
Handled QSFP_DD, QSFP28/QSFP+, SFP+ accordingly based on media type detected.  
#### How to verify it
Check sfpshow eeprom -d, sfpshow presence.
verify whether pmon is running properly and check the syslogs related to pmon docker
UT:
[z9332f_sfp_UT.txt](https://github.com/Azure/sonic-buildimage/files/6389366/z9332f_sfp_UT.txt)

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [x] 202012

### CODE REVIEW

**Comment by aravindmani-1** [comment]
@jleveque , @daall Can you please review the changes?.

**Review by jleveque — APPROVED** [review]

**Comment by aravindmani-1** [comment]
@jleveque https://github.com/Azure/sonic-buildimage/pull/6988 has to be merged first in 202012 branch and then this PR can be merged next, so that the the cherry pick would work properly.

File: platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py [hunk 1]
```diff
@@ -410,58 +502,135 @@ def get_transceiver_info(self):
...
+                #Since the cable length field can be returned as decimal and float in M,converting it to float as common type.
+                #If the currentreturn type persists, cable length has to defined twice for the same length(e.g. 1 and 1.0M)
+                transceiver_info_dict['cable_length'] = str(float(cable_length['data']['Length Cable Assembly(m)']['value']))
+
+            transceiver_info_dict['encoding'] = "Not supported"
+            transceiver_info_dict['nominal_bit_rate'] = "Not supported"
+            transceiver_info_dict['ext_rateselect_compliance'] = "Not supported"
+
+            eeprom_raw = []
+            eeprom_raw = self._read_eeprom_bytes(self.eeprom_path, QSFP_DD_MEDIA_TYPE_OFFSET,\
```
Comment (jleveque) [inline platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py:565]: Trailing backslash should not be necessary due to parentheses.

File: platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py [hunk 2]
```diff
@@ -410,58 +502,135 @@ def get_transceiver_info(self):
...
+                transceiver_info_dict['hardware_rev'] = vendor_rev['data']['Vendor Rev']['value']
+
+            vendor_sn = self._get_eeprom_data('serial')
+            if vendor_sn is not None:
+                transceiver_info_dict['serial'] = vendor_sn['data']['Vendor SN']['value']
+
+            if self.qsfp_dd_Info is None:
+                return None
+
+            sfp_media_type_raw = self._read_eeprom_bytes(self.eeprom_path, QSFP_DD_MEDIA_TYPE_OFFSET,\
```
Comment (jleveque) [inline platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py:599]: Trailing backslash should not be necessary due to parentheses.

File: platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py [hunk 3]
```diff
@@ -410,58 +502,135 @@ def get_transceiver_info(self):
...
+                                                         QSFP_DD_MEDIA_TYPE_WIDTH)
+            if sfp_media_type_raw is not None:
+                sfp_media_type_dict = self.qsfp_dd_Info.parse_media_type(sfp_media_type_raw, 0)
+                if sfp_media_type_dict is None:
+                    return None
+
+            host_media_list = ""
+            if self.qsfp_dd_Info is None:
+                return None
+            qsfp_dd_app1_list = self._read_eeprom_bytes(self.eeprom_path, QSFP_DD_APP1_ADV_OFFSET,\
```
Comment (jleveque) [inline platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py:609]: Trailing backslash should not be necessary due to parentheses.

File: platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py [hunk 4]
```diff
@@ -410,58 +502,135 @@ def get_transceiver_info(self):
...
+                                                                      QSFP_DD_APP1_ADV_WIDTH)
+            if self.qsfp_dd_app2_list:
+                possible_application_count = 15
+                #Additional application advertisement
+                qsfp_dd_app2_list = self._read_eeprom_bytes(self.eeprom_path, QSFP_DD_APP2_ADV_OFFSET,\
```
Comment (jleveque) [inline platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py:614]: Trailing backslash should not be necessary due to parentheses.

File: platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py [hunk 5]
```diff
@@ -410,58 +502,135 @@ def get_transceiver_info(self):
...
-#            ext_media_dict = ext_media_module.get_ext_media_info(self)
-#            for key in ext_media_dict:
-#                value = ext_media_dict[key]
-#                if value in [None, 'None', 'none','n/a', '']:
-#                    value = 'N/A'
-#                transceiver_info_dict[key] = str(value)
+            for i in range(0, possible_application_count):
+                if sfp_application_type_list[i * 4] == 'ff':
+                    break
+                host_electrical, media_interface = self.qsfp_dd_Info.parse_application(sfp_media_type_dict,\
```
Comment (jleveque) [inline platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py:630]: Trailing backslash should not be necessary due to parentheses.

File: platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py [hunk 6]
```diff
@@ -676,10 +901,20 @@ def get_rx_los(self):
         """
         rx_los_list = []
         try:
-            if  self.media_type.startswith('QSFP'):
+            if self.sfp_type == 'QSFP_DD':
+                offset = 512
+                rx_los_mask = [ 0x01, 0x02, 0x04, 0x08 ,0x10, 0x20, 0x40, 0x80 ]
+                dom_channel_monitor_raw = self._read_eeprom_bytes(self.eeprom_path,\
```
Comment (jleveque) [inline platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py:907]: Trailing backslash should not be necessary due to parentheses.

File: platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py [hunk 7]
```diff
@@ -807,14 +1069,29 @@ def get_tx_bias(self):
...
+            offset = 128
+            if self.sfp_type == 'QSFP_DD':
+                if self.qsfp_dd_DomInfo is None:
+                    return None
+                if not self.dom_tx_bias_supported:
+                    return None
+                tx_bias_data_raw = self._read_eeprom_bytes(self.eeprom_path, offset + QSFP_DD_TXBIAS_OFFSET, QSFP_DD_TXBIAS_WIDTH)
+                tx_bias_data = self.qsfp_dd_DomInfo.parse_dom_tx_bias(tx_bias_data_raw, 0)
+
+                for tx_bias_id in ('TX1Bias', 'TX2Bias', 'TX3Bias', 'TX4Bias',\
```
Comment (jleveque) [inline platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py:1081]: Trailing backslash should not be necessary due to parentheses.

File: platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py [hunk 8]
```diff
@@ -824,15 +1101,29 @@ def get_rx_power(self):
...
-                for rx_power_id in ('RX1Power', 'RX2Power', 'RX3Power', 'RX4Power'):
+            if self.sfp_type == 'QSFP_DD':
+                if self.qsfp_dd_DomInfo is None:
+                    return None
+                if not self.dom_rx_power_supported:
+                    return None
+                rx_power_data_raw = self._read_eeprom_bytes(self.eeprom_path, offset + QSFP_DD_RXPOWER_OFFSET, QSFP_DD_TXPOWER_WIDTH)
+                rx_power_data = self.qsfp_dd_DomInfo.parse_dom_rx_power(rx_power_data_raw, 0)
+
+                for rx_power_id in ('RX1Power', 'RX2Power', 'RX3Power', 'RX4Power',\
```
Comment (jleveque) [inline platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py:1114]: Trailing backslash should not be necessary due to parentheses.

File: platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py [hunk 9]
```diff
@@ -842,8 +1133,25 @@ def get_tx_power(self):
...
         try:
-            if self.media_type.startswith('QSFP'):
+            if self.sfp_type == 'QSFP_DD':
+                if self.qsfp_dd_DomInfo is None:
+                    return None
+
+                if not self.dom_tx_power_supported:
+                    return None
+
+                tx_power_data_raw = self._read_eeprom_bytes(self.eeprom_path, offset + QSFP_DD_TXPOWER_OFFSET,\
```
Comment (jleveque) [inline platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py:1145]: Trailing backslash should not be necessary due to parentheses.

File: platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py [hunk 10]
```diff
@@ -842,8 +1133,25 @@ def get_tx_power(self):
...
+                                                         QSFP_DD_TXPOWER_WIDTH)
+                tx_power_data = self.qsfp_dd_DomInfo.parse_dom_tx_power(tx_power_data_raw, 0)
+
+                for tx_power_id in ('TX1Power', 'TX2Power', 'TX3Power', 'TX4Power',\
```
Comment (jleveque) [inline platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py:1149]: Trailing backslash should not be necessary due to parentheses.

File: platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py [hunk 11]
```diff
@@ -1031,17 +1339,20 @@ def reinit_sfp_driver(self):
                 driver_name = driver_name.lstrip(" ")
 
             #Avoid re-initialization of the QSFP/SFP optic on QSFP/SFP port.
-            if (self.media_type == 'SFP' and (driver_name == 'optoe1' or driver_name == 'optoe3')):
+            if self.sfp_type == 'SFP' and\
+                (driver_name == 'optoe1' or driver_name == 'optoe3'):
```
Comment (jleveque) [inline platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py:1343]: Suggest simplifying:

```python
if self.sfp_type == 'SFP' and driver_name in ['optoe1', 'optoe3']):
```

File: platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py [hunk 12]
```diff
@@ -1031,17 +1339,20 @@ def reinit_sfp_driver(self):
...
                 subprocess.Popen(delete_device, shell=True, stdout=subprocess.PIPE)
                 new_device = "echo optoe2 0x50 >" + new_sfp_path
                 subprocess.Popen(new_device, shell=True, stdout=subprocess.PIPE)
                 time.sleep(2)
-            elif (self.media_type == 'QSFP' and (driver_name == 'optoe2' or driver_name  == 'optoe3')):
+            elif self.sfp_type == 'QSFP' and\
+                 (driver_name == 'optoe2' or driver_name  == 'optoe3'):
```
Comment (jleveque) [inline platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py:1349]: Suggest simplifying:

```python
if self.sfp_type == 'QSFP' and driver_name in ['optoe2', 'optoe3']):
```

File: platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py [hunk 13]
```diff
@@ -1031,17 +1339,20 @@ def reinit_sfp_driver(self):
...
                 subprocess.Popen(delete_device, shell=True, stdout=subprocess.PIPE)
                 new_device = "echo optoe1 0x50 >" + new_sfp_path
                 subprocess.Popen(new_device, shell=True, stdout=subprocess.PIPE)
                 time.sleep(2)
-            elif (self.media_type == 'QSFP-DD' and (driver_name == 'optoe1' or driver_name  == 'optoe2')):
+            elif self.sfp_type == 'QSFP_DD' and\
+                 (driver_name == 'optoe1' or driver_name  == 'optoe2'):
```
Comment (jleveque) [inline platform/broadcom/sonic-platform-modules-dell/z9332f/sonic_platform/sfp.py:1355]: Suggest simplifying:

```python
if self.sfp_type == 'QSFP_DD' and driver_name in ['optoe1', 'optoe2']):
```

**Review by jleveque — CHANGES_REQUESTED** [review]

**Review by jleveque — APPROVED** [review]

**Pull request merged**